dask/diagnostics/profile.py

```diff
@@ -84,10 +90,124 @@ def visualize(self, **kwargs):
         dask.diagnostics.profile_visualize.visualize
         """
         from .profile_visualize import visualize
-        return visualize(self.results, self._dsk, **kwargs)
+        return visualize(self, **kwargs)
 
     def clear(self):
         """Clear out old results from profiler"""
         self._results.clear()
         del self.results[:]
         self._dsk = {}
+
+
+ResourceData = namedtuple('ResourceData', ('time', 'mem', 'cpu'))
+
+
+class ResourceProfiler(Callback):
+    """A profiler for resource use.
+
+    Records the following each timestep
+        1. Time in seconds since the epoch
+        2. Memory usage in MB
+        3. % CPU usage
+
+    Examples
+    --------
+
+    >>> from operator import add, mul
+    >>> from dask.threaded import get
+    >>> dsk = {'x': 1, 'y': (add, 'x', 10), 'z': (mul, 'y', 2)}
+    >>> with ResourceProfiler() as prof:  # doctest: +SKIP
+    ...     get(dsk, 'z')
+    22
+
+    These results can be visualized in a bokeh plot using the ``visualize``
+    method. Note that this requires bokeh to be installed.
+
+    >>> prof.visualize() # doctest: +SKIP
+
+    You can activate the profiler globally
+
+    >>> prof.register()  # doctest: +SKIP
+
+    If you use the profiler globally you will need to clear out old results
+    manually.
+
+    >>> prof.clear()
+    """
+    def __init__(self, dt=1):
+        self._tracker = _Tracker(dt)
+        self._tracker.start()
+        self.results = []
+
+    def _start(self, dsk):
+        assert self._tracker.is_alive(), "Resource tracker is shutdown"
+        self._tracker._trigger.set()
+
+    def _finish(self, dsk, state, failed):
+        self._tracker._trigger.clear()
+        self.results.extend(starmap(ResourceData, self._tracker.queue.get()))
+
+    def close(self):
+        """Shutdown the resource tracker process"""
+        self._tracker.shutdown()
+
+    def clear(self):
+        self.results = []
+
+    def _plot(self, **kwargs):
+        from .profile_visualize import plot_resources
+        return plot_resources(self.results, **kwargs)
+
+    def visualize(self, **kwargs):
+        """Visualize the profiling run in a bokeh plot.
+
+        See also
+        --------
+        dask.diagnostics.profile_visualize.visualize
+        """
+        from .profile_visualize import visualize
+        return visualize(self, **kwargs)
+
+
+class _Tracker(Process):
+    """Background process for tracking resource usage"""
+    def __init__(self, dt=1):
+        import psutil
+        Process.__init__(self)
+        self.daemon = True
+        self.dt = dt
+        self.parent = psutil.Process(current_process().pid)
+        self.queue = Queue()
+        self._trigger = Event()     # Sleeps until trigger
+        self._shutdown = Event()    # Set when process should shutdown
+
+    def shutdown(self):
+        self._shutdown.set()
+        self._trigger.set()
+        self.queue.close()
+        self.join()
+
+    __del__ = shutdown
```

Comment:
Does it work when the object is cleaned up not during an exit?  

Strange things happen during exit.  It's hard to guarantee that the right bits of code will be around as the interpreter shuts down.